### DISCUSSION TITLE
"stg push" not indicating if the patch was modified

### DISCUSSION THREAD

**Discussion by shivaprasadbhat** [stacked-git/stgit#464] [Q&A] [answered]
With recent update, if the patch is modified automatically when doing "stg push" Ex: Few hunks of the patch up on the stack are copied to the patches on the bottom of stack. 

Earlier "stg push" used to mention "modified". Now, it doesnt! Is it expected?

**Comment by jpgrayson** [comment]
I think you are correct that there has been a regression w.r.t. `stg push` not reporting when pushed patches have been modified.

**Reply by jpgrayson** [reply, marked as answer]
Fixed in 2868ffe0458a3578e43dcf1edd84b9c3e9e671c3.

**Reply by shivaprasadbhat** [reply]
Excellent!! Thank you!

**Reply by jpgrayson** [reply]
I may have been too hasty with the fix for this that went into StGit 2.4.11. I think StGit may have gone from under-reporting modified patches to now _over_-reporting.

Specifically, the new behavior is for "(modified)" to be reported anytime the patch's commit changes. This is much more aggressive than StGit 1.x where "(modified)" would only be reported if the apparent diff of a patch changed. I.e. only reporting "(modified)" if hunks were added, removed, or moved.

StGit 1.x and 2.x have some differences in the exact git/gix operations performed when pushing patches. I need to look at this again, but IIRC the git commands StGit 1.x used provided a clear signal about whether a patch was "modified" in the limited "diff changed" sense whereas that signal is not as readily determined in StGit 2.x (but StGit 2.x does something faster).

Any feedback about new/old/ideal behavior(s) in this space would be appreciated.

**Reply by jpgrayson** [reply]
Okay, what's going on is that in StGit 2.0 we started using the `--3way` option to `git apply` when pushing patches which generally allows patch application to succeed first-try whereas in StGit 1.0 patches that required any sort of non-trivial patch application would first fail with `git apply` and then fallback to using `git merge-recursive`. Whenever the fallback to `git merge-recursive` was needed, that was when the "(modified)" status would be shown.

So in StGit 2.0, it is still conceivable that some patches might hit the `git merge-recursive` fallback and thus _could_ show up as "(modified)", but in practice it doesn't happen.

The thing about the current patch application flow using `git apply --3way` is that it is fast and correct. So reverting back just so we can see "(modified)" doesn't feel like the right move to me.

The change I'm planning on making is to add a new "(updated)" status that indicates when a patch's commit is changed. I.e. replacing the overzealous and misleading "(modified)" status that went into 2.4.11.

**Reply by shivaprasadbhat** [reply]
Thanks for continuing. I havent had time to try your fix and give feedback. 

I am confused.
- If on StGit 2.0, "git merge-recursive" fallback doesnt happen practically, then isn't the lost (modified) a lost functionality from StGit 1.0 ? 
- Or We will have both (updated) and (modified) shown as and when necessary?(This would be best!)

**Reply by jpgrayson** [reply]
> If on StGit 2.0, "git merge-recursive" fallback doesnt happen practically, then isn't the lost (modified) a lost functionality from StGit 1.0 ?

Yes, it's a regression, but there are tradeoffs of speed and correctness to get a good signal about whether a patch was "modified" during push.

> Or We will have both (updated) and (modified) shown as and when necessary?(This would be best!)

It is _possible_ that there are scenarios where `git apply --3way` is not able to push a patch, but `git merge-recursive` succeeds. In such a case the "(modified)" status would appear. But I don't know how to construct such a scenario. I think it's more likely that either `git apply --3way` succeeds or there will be conflicts that require manual resolution.

-----

Another thing here is that this definition of a patch being "modified" by a push is a bit fuzzy. Some possible ways to define it:

* Plain `git apply` didn't work, but `git merge-recursive` did. (The StGit 1.x definition)
* The post-push diff is not identical to the pre-push diff.
* The post-push diff is _materially_ different than the pre-push diff.
  - Material: a hunk is added or removed
  - Non-material: a hunk is shifted by some number of lines
  - ???: a hunk makes the same change, but its surrounding context has changed
* The set of blobs affected by the post-push diff is different than the set of pre-push blobs OR the sets are the same, but the blob hashes (content) is different.

The reason I'm currently okay with living with this regression is because making any of those determinations is computationally expensive...unless you are making the determination _while_ performing the 3-way patch application. But I don't currently have a way to tell `git apply` to give me side-channel information about how hard it had to try to do its job. And gix does not yet have patch application capabilities.
